### PR TITLE
dev/core#2968 - Fix invalid assets url paths

### PR DESCRIPTION
**Pull request by adriano66**
Overview
----------------------------------------
This fixes assets URLs generation on instances that are using symlinks in the configuration.
I've described it all in the gitlab ticket https://lab.civicrm.org/dev/core/-/issues/2968

### CODE REVIEW

**Comment by colemanw** [comment]
@civicrm-builder add to whitelist

**Comment by eileenmcnaughton** [comment]
@totten @seamuslee001 ?

**Comment by jaapjansma** [comment]
test this please

**Comment by jaapjansma** [comment]
We (@bettydolfing, @kainuk and I) are doing the review of PR's. @colemanw asked us to review this PR however we could not reproduce this problem in https://dmaster.demo.civicrm.org So we cannot verify whether this patch will fix the problem or not. 

So we think if the unit tests does not fail this patch might be worth to merge as it is only a small change in code.

**Comment by adriano66** [comment]
Thanks, like I've explained in the gitlab ticket this issue is only affecting the symlinked paths.

I'm not sure about the exact change that created this issue in one of the versions between 5.12 and 5.42 but updating this path resolver function seems like a safe bet to fix this issue in general.

It's solved in a similar manner in `\CRM\Utils\System\Wordpress.php`, 
Somehow there is no separate path resolver for Drupal so the base one is used.

**Comment by mlutfy** [comment]
jenkins, test this please

**Comment by seamuslee001** [comment]
I think this needs some carful review on the various different platforms, file paths has been one thing that has been issues previously

**Comment by demeritcowboy** [comment]
This doesn't seem to affect windows for drupal 7 and 9, despite the function outputting nonsense both before and after the patch. I can't offhand explain why it all still works, but for what it's worth I don't see any issue on windows that wasn't there before.

File: CRM/Utils/System/Base.php [hunk 1]
```diff
@@ -718,7 +718,8 @@ public function getCiviSourceStorage() {
       // the system for a loop on lobo's macosx box
       // or in modules
       $cmsPath = $config->userSystem->cmsRootPath();
-      $userFrameworkResourceURL = $baseURL . str_replace("$cmsPath/", '',
+      $realCmsPath = realpath($cmsPath);
+      $userFrameworkResourceURL = $baseURL . str_replace(["$cmsPath/", "$realCmsPath/"], '',
```
Comment (pgarbarczyk) [inline CRM/Utils/System/Base.php:722]:
```suggestion
      $realCmsPath = realpath($cmsPath);

      if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
        $cmsPath = strtolower(str_replace('\\', '/', $cmsPath));
        $realCmsPath = strtolower(str_replace('\\', '/', $realCmsPath));
      }

      $userFrameworkResourceURL = $baseURL . str_replace(["$cmsPath/", "$realCmsPath/"], ['', ''],
```

Comment (pgarbarczyk) [reply CRM/Utils/System/Base.php:722]: The fix has a significant impact under Windows. For example, without it, URLs to JS resources have the format `http: //myproject.local/c:/projects/myproject/sites/all/modules/civicrm/packages/jquery/plugins/jquery.timeentry.min.js`.
Below debug of key variables, please pay attention to slash, backslahe, upper and lower case (hard drive letter):
```
$civicrm_root => c:/projects/myproject/sites/all/modules/civicrm/
$cmsPath => C:/xampp-702/htdocs/myproject
$realCmsPath => C:\projects\myproject
```

**Comment by adriano66** [comment]
"Feel free to reopen" - I'm not sure if this message was intended for me but I don't think I'm able to reopen the PR.

This is the real issue we are experiencing that we've patched on our environments and would like this to also be patched in the core so we don't need to patch the CiviCRM after each upgrade.

Can you tell me what exactly needs to be fixed for this to be acceptable?

I didn't want to refactor this as I'm not working with the CiviCRM core daily and honestly don't really have time for that.
But if you tell me how you'd like this to be fixed I'll happily work on that.

The current code takes a very naive approach and ignores the existence of the symlinks altogether. 
Someone, somewhere probably removed some kind of path normalization and here we are with this bug.
To find the root cause of this regression in a sane way we'd need to run bisect on this repository and that requires building CiviCRM for Drupal locally for this to work, which is a pain for me right now, if it even is possible to run git bisect on the built drupal/civicrm instance.

Do I need to create another PR for someone to look at this?

**Comment by mattwire** [comment]
Hi @adriano66 we can easily end up with unreviewable PRs in the queue and closing them after inactivity is a way of reducing the size of the queue. The lab issue still tracks the problem.
However, the problem with closing is that you don't have permission to reopen unless you have extra privileges on github (@demeritcowboy for most users it's not just after they push commits).

I've reopened it for you so tests can run etc. The problem with reviewing this is that it has the potential to affect a lot of different scenarios / platforms. It would also be helpful if you can identify existing test coverage and possibly add additional tests.

**Comment by adriano66** [comment]
@mattwire Thanks for the explanation

**Comment by eileenmcnaughton** [comment]
@totten  I feel like you have dug into this stuff & found problems with similar patches before (ie switching to use realpath) & since this is not on the first page it is almost certainly not on your radar....

**Comment by jaapjansma** [comment]
@eileenmcnaughton or @totten can you give @adriano66 feedback on what he needs to do to get this merged into core. 
It looks like there is one issue with this patch applied when CiviCRM is running on a Windows machine.

**Pull request closed without merging**